### PR TITLE
Improved lineage graph, rewrote History module

### PR DESCRIPTION
**Pull request by sdorkenw**
This PR fixes issues with the lineage graph and improves its performance dramatically by removing redundant calls caused by diverging-converging paths in the lineage. Based on the multi-root lineage, I rewrote the history module and implemented the History object for many-roots. At the moment, only the past-future mapping for the live query is making use of that. We should expose this functionality via other endpoint as well.

### CODE REVIEW

File: pychunkedgraph/backend/history.py [hunk 1]
```diff
@@ -2,72 +2,120 @@
...
 from pychunkedgraph.backend import chunkedgraph_exceptions as cg_exceptions
 
-former_parent_col = column_keys.Hierarchy.FormerParent
-operation_id_col = column_keys.OperationLogs.OperationID
 
-class SegmentHistory(object):
-    def __init__(self, cg, root_id):
-        if not cg.is_root(root_id):
-            raise cg_exceptions.Forbidden
+def get_username_dict(user_ids, auth_token):
```
Comment (akhileshh) [inline pychunkedgraph/backend/history.py:15]: Looks good except this part. I think it is better to pass return value of this function to `History` object. Frontend `app` module is dependent on `backend`, `get_username_dict` calling app server endpoints from inside `backend` feels like a bad circular dependency design even though it works.

`backend` components ideally should not depend on auth or user management system.

Comment (sdorkenw) [reply pychunkedgraph/backend/history.py]: Agreed. Fixed that

**Review by akhileshh — APPROVED** [review]

File: pychunkedgraph/app/segmentation/common.py [hunk 2]
```diff
@@ -922,15 +931,12 @@ def _parse_timestamp(arg_name, default_timestamp=0):
...
-        graph = lineage_graph(cg, root_ids, timestamp_past, timestamp_future)
+        graph = lineage.lineage_graph(cg, root_ids, timestamp_past, timestamp_future)
         return node_link_data(graph)
-    graph = cg_history.SegmentHistory(cg, int(root_id)).get_change_log_graph(
-        timestamp_past, timestamp_future
-    )
-    return node_link_data(graph)
+
+    history_ids = cg_history.History(cg, int(root_id), timestamp_past, timestamp_future)
+    return node_link_data(history.lineage_graph)
```
Comment (fcollman) [inline pychunkedgraph/app/segmentation/common.py:939]: typo here... should be history_ids

**Comment by fcollman** [comment]
I've now tested the id_mapping feature of this (which is how i found the bugs my commits squash).

File: pychunkedgraph/app/segmentation/common.py (hunk 2 above)
Comment (sdorkenw) [reply pychunkedgraph/app/segmentation/common.py]: That does not look right to me...

**Pull request merged**